File: solutions/Golang/median_sorted_arrays_binary_search.go
package main

import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    if len(nums1) > len(nums2) {
        return findMedianSortedArrays(nums2, nums1)
    }

    m, n := len(nums1), len(nums2)
    left, right := 0, m

    for left <= right {
        partition1 := (left + right) / 2
        partition2 := (m + n + 1) / 2 - partition1

        left_max1 := math.MinInt32
        if partition1 != 0 { left_max1 = nums1[partition1-1] }
        right_min1 := math.MaxInt32
        if partition1 != m { right_min1 = nums1[partition1] }

        left_max2 := math.MinInt32
        if partition2 != 0 { left_max2 = nums2[partition2-1] }
        right_min2 := math.MaxInt32
        if partition2 != n { right_min2 = nums2[partition2] }

        if left_max1 <= right_min2 && left_max2 <= right_min1 {
            if (m+n)%2 == 0 {
                return float64(max(left_max1, left_max2)+min(right_min1, right_min2)) / 2.0
            } else {
                return float64(max(left_max1, left_max2))
            }
        } else if left_max1 > right_min2 {
            right = partition1 - 1
        } else {
            left = partition1 + 1
        }
    }

    return -1.0
}

func max(a, b int) int { if a > b { return a }; return b }
func min(a, b int) int { if a < b { return a }; return b }

File: solutions/JavaScript/median_sorted_arrays_binary_search.js
function findMedianSortedArrays(nums1, nums2) {
    if (nums1.length > nums2.length) {
        return findMedianSortedArrays(nums2, nums1);
    }

    const m = nums1.length, n = nums2.length;
    let left = 0, right = m;

    while (left <= right) {
        const partition1 = Math.floor((left + right) / 2);
        const partition2 = Math.floor((m + n + 1) / 2) - partition1;

        const left_max1 = partition1 === 0 ? -Infinity : nums1[partition1 - 1];
        const right_min1 = partition1 === m ? Infinity : nums1[partition1];
        const left_max2 = partition2 === 0 ? -Infinity : nums2[partition2 - 1];
        const right_min2 = partition2 === n ? Infinity : nums2[partition2];

        if (left_max1 <= right_min2 && left_max2 <= right_min1) {
            return (m + n) % 2 === 0 ?
                (Math.max(left_max1, left_max2) + Math.min(right_min1, right_min2)) / 2 :
                Math.max(left_max1, left_max2);
        } else if (left_max1 > right_min2) {
            right = partition1 - 1;
        } else {
            left = partition1 + 1;
        }
    }

    return -1.0;
}
